### PR TITLE
[JENKINS-56477] Fix to allow selecting compatible-only updates for plugins

### PR DESCRIPTION
**Pull request by romenrg**
See [JENKINS-56477](https://issues.jenkins-ci.org/browse/JENKINS-56477).

In Nov 2018 [this PR](https://github.com/jenkinsci/jenkins/pull/3715) was approved and merged, but later @daniel-beck noticed there was a bug in the selection of "compatible plugins", which was not taking all cases into account.

In this PR I have modified the condition to select the compatible plugins, so that it covers all cases and, following @oleg-nenashev suggestion, I've created a separate button to allow users to install either "All", "Compatible" or "None".

### Proposed changelog entries

* The Plugin Manager now provides improved selection for plugin updates. Users willing to update plugins can select "All", "Compatible" or "None".
* The selection of "Compatible" plugins has been fixed to exclude all that contain any warnings

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck - As the original reporter of the ticket
@oleg-nenashev - Your [suggestion to provide a third option for the selection](https://github.com/jenkinsci/jenkins/pull/3715#pullrequestreview-170569822) was taken into account
@batmat - I followed your TO-DO in the UpdateSiteTest and factored out the initialization piece for the UpdateSite creation that was being used in the "minimumJavaVersion" test and is now also included in the test I added (isPluginCompatible). Notice I removed the "sites" variable, since it seemed to be unused; using just the specific site.

### CODE REVIEW

**Comment by romenrg** [comment]
I'm attaching here a screenshot of the result of clicking the new "Compatible" link at the button of the Plugin Manager page:

fixed-select-compatible

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateSite.java]: Cannot be exported due to argument, so should be throwing an exception in the remote API (or at least log a warning that this is un-exportable)

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateSite.java]: Unnecessary?

File: core/src/main/resources/hudson/PluginManager/_table.js
Comment (daniel-beck) [inline core/src/main/resources/hudson/PluginManager/_table.js]: `inputs[i]` == `candidate` so why are we not using the nicer local var here? Especially since with the next line this is really confusing code.

Also, this can probably be simplified to:

         if(candidate.type === "checkbox") {
            candidate.checked = candidate.dataset.compatWarning === 'false';
        }

File: core/src/test/java/hudson/model/PluginTest.java
Comment (daniel-beck) [inline core/src/test/java/hudson/model/PluginTest.java]: `@Issue("JENKINS-56477")`

File: core/src/test/java/hudson/model/PluginTest.java
Comment (daniel-beck) [inline core/src/test/java/hudson/model/PluginTest.java]: What does the use of Mockito here gain you? Doesn't the other (modified) test not show this is unnecessary?

Comment (romenrg) [reply core/src/test/java/hudson/model/PluginTest.java]: Yes, I've removed this unit test and I'm now leaving only the other integration test as sufficient.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: I'm not sure how the @Exported annotation is being used. I included it in both cases as I saw it in the original "isCompatibleWithInstalledVersion", but now that I'm looking to it in more detail, I assume it is only used as a way to publish the method as an API operation, right?

If that is the case, it seems that for this method overload the annotation is not necessary at all.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: I assume the fact that there is a jelly file that invokes it doesn't imply that the method has to be annotated with "Exported", right? I wasn't sure how this was being used.

If that is not a requirement, then there is no need to have it, so I'll be happy to remove it.

File: core/src/main/resources/hudson/PluginManager/_table.js
Comment (romenrg) [reply core/src/main/resources/hudson/PluginManager/_table.js]: You are right, should be using `candidate` instead.

That said, I really like your proposed alternative. Very elegant. Will use it :)

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: Removed and tested locally afterwards, working fine.

File: core/src/main/resources/hudson/PluginManager/_table.js
Comment (romenrg) [reply core/src/main/resources/hudson/PluginManager/_table.js]: Done 👍

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateSite.java]: Would `@Restricted(NoExternalUse.class)` this, no reason IMO to expose this to plugins.

If we end up not doing that, needs Javadoc + `@since TODO`.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/UpdateSite.java]: Would `@Restricted(NoExternalUse.class)` this, no reason IMO to expose this to plugins.

If we end up not doing that, needs Javadoc + `@since TODO`.

**Review by daniel-beck — COMMENTED** [review]
Minor remarks, else LGTM.

**Review by vilacides — APPROVED** [review]

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: I've restricted this one, since in case some plugins want to call this, it seems they should be calling the version without the "cache" param, as is the pattern for similar methods.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: For this case I've included the Javadoc and the `@since TODO`.

I might be wrong, but I see this new method as the replacement of the one that was being called previously (`isCompatibleWithInstalledVersion`), which is being exported; so I thought it could be useful to export this one so that whoever uses the "old" one can switch to this new and more comprehensive check

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (daniel-beck) [reply core/src/main/java/hudson/model/UpdateSite.java]: This is for the "REST" API, i.e. `/pluginManager/api` or `/updateCenter/api`, don't remember which, and entirely optional.

Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: Thanks for the explanation, Daniel 👍

Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: Since there have been no comments from others on this, I've applied @daniel-beck 's suggestion and removed the `@Exported` annotation. Definitely not necessary for now.

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (romenrg) [reply core/src/main/java/hudson/model/UpdateSite.java]: Since nobody else has commented on this, I've applied @daniel-beck 's suggestion and made the `isCompatible` methid restricted (`@Restricted(NoExternalUse.class)`).

Definitely not necessary to expose it to plugins for now. And if in the future someone sees value in calling it from a plugin this can always be changed.

**Comment by romenrg** [comment]
In the absence of other comments, I've gone ahead and apply @daniel-beck 's suggestions, so that the new method is not exported nor exposed to plugins. In case in future someone sees value on using it from a plugin, it can always be changed.

Thanks Daniel for your comments and explanations :)

File: core/src/main/resources/hudson/PluginManager/index.jelly
Comment (oleg-nenashev) [inline core/src/main/resources/hudson/PluginManager/index.jelly]: I would rather put "Compatible" before "All".

**Review by oleg-nenashev — APPROVED** [review]
LGTM, it will definitely improve the situation.
Not sure how ATH tests would react to it tho

**Comment by oleg-nenashev** [comment]
@daniel-beck WDYT?

**Comment by oleg-nenashev** [comment]
On hold for this weekly to let other reviewers to provide feedback.
Will merge it if @daniel-beck or @batmat bless it

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> let other reviewers to provide feedback

This is two weeks old. People had a chance to review.

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> WDYT?

I stand by https://github.com/jenkinsci/jenkins/pull/3985#pullrequestreview-226401995. The minor remarks have since been addressed.

**Review by daniel-beck — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
🚢 🇮🇹 then. Thanks @romenrg !

**Pull request merged**
